src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs

```diff
@@ -218,93 +221,233 @@ public static unsafe int IndexOf(ref char searchSpace, char value, int length)
-                    length = (Vector<ushort>.Count - unaligned) & (Vector<ushort>.Count - 1);
+                    lengthToExamine = UnalignedCountVector128(ref searchSpace);
                 }
-
-            SequentialScan:
-                while (length >= 4)
+            }
+            else if (Vector.IsHardwareAccelerated)
+            {
+                // Needs to be double length to allow us to align the data first.
+                if (length >= Vector<ushort>.Count * 2)
                 {
-                    length -= 4;
+                    lengthToExamine = UnalignedCountVector(ref searchSpace);
+                }
+            }
 
-                    if (pCh[0] == value)
-                        goto Found;
-                    if (pCh[1] == value)
-                        goto Found1;
-                    if (pCh[2] == value)
-                        goto Found2;
-                    if (pCh[3] == value)
-                        goto Found3;
+        SequentialScan:
+            // In the non-vector case lengthToExamine is the total length.
+            // In the vector case lengthToExamine first aligns to Vector,
+            // then in a second pass after the Vector lengths is the 
+            // remaining data that is shorter than a Vector length.
+            while (lengthToExamine >= 4)
+            {
+                ref char current = ref Add(ref searchSpace, offset);
+
+                if (value == current)
+                    goto Found;
+                if (value == Add(ref current, 1))
+                    goto Found1;
+                if (value == Add(ref current, 2))
+                    goto Found2;
+                if (value == Add(ref current, 3))
+                    goto Found3;
+
+                offset += 4;
+                lengthToExamine -= 4;
+            }
 
-                    pCh += 4;
-                }
+            while (lengthToExamine > 0)
+            {
+                if (value == Add(ref searchSpace, offset))
+                    goto Found;
 
-                while (length > 0)
+                offset += 1;
+                lengthToExamine -= 1;
+            }
+
+            // We get past SequentialScan only if IsHardwareAccelerated or intrinsic .IsSupported is true. However, we still have the redundant check to allow
+            // the JIT to see that the code is unreachable and eliminate it when the platform does not have hardware accelerated.
+            if (Avx2.IsSupported)
+            {
+                if (offset < length)
                 {
-                    length--;
+                    Debug.Assert(length - offset >= Vector128<ushort>.Count);
+                    if ((((nint)Unsafe.AsPointer(ref searchSpace) + (nint)offset) & (nint)(Vector256<ushort>.Count - 1)) != 0)
+                    {
+                        // Not currently aligned to Vector256 (is aligned to Vector128); this can cause a problem for searches
```

Comment:
I think the comment could be clarified to explain the logic here.

Basically, under optimal conditions we are not 32-byte aligned and are instead 16-byte aligned. In the non-optimal case, we could have any alignment, but we definitely have at least 16-bytes available to read, so we shouldn't fault.

Reply:
However, I'm not sure what (in the non-optimal case) prevents an AV for searches with no upper bound (String.wcslen)....

That is, if someone creates a span over unpinned data and with `length: int.MaxValue` and then they call `IndexOf('\0')`, the length could be less than 16, and it could be relocated to just before the end of a page, which could then fault when we do the `LoadVector128` below, right?

Reply:
> However, I'm not sure what (in the non-optimal case) prevents an AV for searches with no upper bound (String.wcslen)....

`String.wcslen(char* ptr)` is over pointer data so the input is fixed data and the GC can't relocate it. No guarantees are offered for unfixed data and if you are creating a Span that is larger than data you own you're already in trouble; if there is no null terminator then it will rightfully fault.

Reply:
We should likely have a comment explicitly calling that out then. Specifically that this assumes that the length is either correct or that the data is pinned and that you may AV otherwise.

Reply:
Improved comment